### DISCUSSION TITLE
Support for GraphQL persisted queries

### DISCUSSION THREAD

**Discussion by Mosnar** [craftcms/cms#8964] [Ideas]
### Description

The concept of persisted queries in GraphQL is essentially that you keep track of your queries in an allow-list with queries mapped to IDs. When executing the query, instead of passing the entire thing, you just pass a `queryId`. The GraphQL server is then responsible for looking up that query and ensuring it's allowed. Generally, the process of generating the allow list is automated using a frontend build tool to extract queries and generate the allow-list for you, replacing queries with their IDs in your compiled code. 

The feature already exists on the webonyx/graphql-php server via the `persistentQueryLoader` config; however, it would need to be implemented in Craft's implementation.

### Benefits
- Makes it easier to utilize GraphQL as an API in production since you don't need to worry about permissions leaking data
- Smaller client requests since the payload only needs to carry the query ID
- Smaller production assets since your queries would be replaced with IDs
- Improves cachability of http requests since not all full-length queries can fit into GET request parameters to be cached on CDNs

### Related Reading
Relay Persisted Queries Reference: https://relay.dev/docs/en/persisted-queries.html
Apollo Server Reference: https://www.apollographql.com/docs/apollo-server/performance/apq/
Example Webpack Loader: https://github.com/leoasis/graphql-persisted-document-loader

**Comment by denisyilmaz** [comment]
Is there any plan to implement this feature?

**Reply by andris-sevcenko** [reply]
I'd like to add this for Craft 4.0!

**Reply by denisyilmaz** [reply]
Is this still coming for v4?

we are in the planning the next months for existing project in regards to updates, performance enhancements etc. So it would be great to know if this is something that we can expect this year.

**Reply by andris-sevcenko** [reply]
It might not happen for 4.0.0 (depending on how far we get with more crucial things), but it's definitely something I will add for v4.

**Reply by denisyilmaz** [reply]
any update whether this will land in 4.0.0?

**Reply by andris-sevcenko** [reply]
4.0.0 got a feature freeze as soon as the beta was launched. Unfortunately, this didn't make the cut, sorry. Still possible for 4.1!

**Reply by denisyilmaz** [reply]
Hi @andris-sevcenko any updates on this feature?

**Reply by mark-wff** [reply]
@brandonkelly is there an update when the feature will be implemented?

**Reply by denisyilmaz** [reply]
@brandonkelly any updates regarding this feature?

**Reply by brandonkelly** [reply]
We’re planning to do a big GraphQL push soon, once the dust has cleared from the Craft 5 launch!

**Reply by denisyilmaz** [reply]
That sounds great. Using Craft CMS in Headless mode is our default setup for clients, so i'm happy to see progress in this area
.

**Comment by denisyilmaz** [comment]
Hi @andris-sevcenko besides the idea of persisted queries, would something like "RESTified GraphQL Endpoints" be easier to implement? Hasura GraphQL Engine is using this idea: https://hasura.io/docs/latest/api-reference/restified/
You basically save GraphQL queries in the database and create a RestAPI endpoint for it.

**Reply by brandonkelly** [reply]
Can you post that as a new discussion?

**Reply by denisyilmaz** [reply]
created the discussion here: https://github.com/craftcms/cms/discussions/12503

**Comment by denisyilmaz** [comment]
@brandonkelly are there any plans to integrate this? maybe for craft 6?

For bigger projects we see GraphQL being the biggest issue in regards to performance when the IA is using a lot of relationship fields and the content gets updated regularly. any update for a stronger/better caching/performance for graphql would highly appreciated

**Reply by denisyilmaz** [reply]
@brandonkelly @andris-sevcenko Any update on this?

**Reply by brandonkelly** [reply]
Sorry, no update yet, but it’s on our list.
